services/streammanager/kafka/client/producer.go

```diff
@@ -117,5 +131,20 @@ func (p *Producer) Publish(ctx context.Context, msgs ...Message) error {
 			Headers: headers,
 		}
 	}
-	return p.writer.WriteMessages(ctx, messages...)
+
+	var err error
+	for i := 0; i < p.config.MaxRetries; i++ {
+		err = p.writer.WriteMessages(ctx, messages...)
+		if errors.Is(err, kafka.LeaderNotAvailable) {
```

Comment:
We only sleep if the error is `kafka.LeaderNotAvailable`, why not sleeping in other cases as well ?

Reply:
@lvrach I simply got this code from their docs (they retry on this error only or on `context.DeadlineExceeded`) but you are right, it can make sense to retry on other errors too.

From what I can see, errors that can make sense retry might be:
* NetworkException
* NotEnoughReplicas
* NotEnoughReplicasAfterAppend
* NotController
* KafkaStorageError
* FencedLeaderEpoch
* UnknownLeaderEpoch
* PreferredLeaderNotAvailable
* EligibleLeadersNotAvailable
* RequestTimedOut
* ListenerNotFound
* FetchSessionIDNotFound
* InvalidFetchSessionEpoch
* ElectionNotNeeded

In the segment lib they added a `Temporary() bool` method to errors which returns true
if those ops can be retried.

https://github.com/segmentio/kafka-go/blob/0b5e1f680537e6d0b6b754601f808e366f7df2ba/error.go#L138

All the errors I mentioned above are in the `Temporary() bool` so we could use that
but please consider that that method contains also non-producer errors and two errors
that are not retryable, at least according to the Kafka documentation:
* NotLeaderForPartition <- not retryable
* NoReassignmentInProgress <- not retryable

We could:
1. not care that `Temporary() bool` also has errors not returned by producers, open
   an issue on Segment asking for those two errors that can't be retried and simply
   use `if err.Temporary()` to determine whether we can retry the message
2. build our own `Temp` method

I'd say 1 but it requires more time. WDYT?

Reply:
In the context of the `Producer.Publish` method I would say we shouldn't care about retries. Rudder-server will take care of aborting or retrying if necessary. Regarding whether an error is abortable or not, we need to honour what we are currently doing in `kafkaerror.go`

Reply:
It is done internally already by the `kafka.Writer`, I set up the `MaxAttempts` and it is working. If you guys are OK with that please resolve.